server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
+                    } catch (Exception ex) {
                         logger.warn(() ->
                             new ParameterizedMessage("[{}] [{}] failed to read metadata for index", snapshotId, index), ex);
                     }
-
                     deleteIndexMetaDataBlobIgnoringErrors(snapshot, indexId);
-
                     if (indexMetaData != null) {
                         for (int shardId = 0; shardId < indexMetaData.getNumberOfShards(); shardId++) {
                             try {
-                                delete(snapshotId, indexId, new ShardId(indexMetaData.getIndex(), shardId));
+                                final ShardId sid = new ShardId(indexMetaData.getIndex(), shardId);
+                                new Context(snapshotId, indexId, sid, sid).delete();
                             } catch (SnapshotException ex) {
                                 final int finalShardId = shardId;
                                 logger.warn(() -> new ParameterizedMessage("[{}] failed to delete shard data for shard [{}][{}]",
                                     snapshotId, index, finalShardId), ex);
                             }
                         }
                     }
-                }
+                    if (outstandingIndices.decrementAndGet() == 0) {
+                        deleteUnreferencedIndices(repositoryData, updatedRepositoryData, listener);
+                    }
+                });
             }
+        } else {
+            deleteUnreferencedIndices(repositoryData, updatedRepositoryData, listener);
+        }
+    }
 
-            // cleanup indices that are no longer part of the repository
-            final Collection<IndexId> indicesToCleanUp = Sets.newHashSet(repositoryData.getIndices().values());
-            indicesToCleanUp.removeAll(updatedRepositoryData.getIndices().values());
-            final BlobContainer indicesBlobContainer = blobStore().blobContainer(basePath().add("indices"));
-            for (final IndexId indexId : indicesToCleanUp) {
+    private void deleteUnreferencedIndices(RepositoryData repositoryData, RepositoryData updatedRepositoryData,
+            ActionListener<Void> listener) {
+        // cleanup indices that are no longer part of the repository
+        final Collection<IndexId> indicesToCleanUp = Sets.newHashSet(repositoryData.getIndices().values());
+        indicesToCleanUp.removeAll(updatedRepositoryData.getIndices().values());
+        final BlobContainer indicesBlobContainer = blobStore().blobContainer(basePath().add("indices"));
+        if (indicesToCleanUp.isEmpty()) {
+            // We're done, no indices to clean up
+            listener.onResponse(null);
+            return;
+        }
+        final AtomicInteger outstanding = new AtomicInteger(indicesToCleanUp.size());
+        for (final IndexId indexId : indicesToCleanUp) {
+            threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(() -> {
```

Comment:
ActionRunnable and GroupedListner

Reply:
Used `GroupedListner` but see below, I think we can do without `ActionRunnable` and keep things a little simpler.

Reply:
Right, this spot fortunately went away on its own now due to having bulk deletes :)